theano/tensor/signal/downsample.py

```diff
@@ -882,6 +878,178 @@ def grad(self, inp, grads):
+            raise NotImplementedError() #fallback to our perform()
+        x, gz = inp
+        gx, = out
+        fail = sub['fail']        
+        ignore_border = int(self.ignore_border)
+        ds0, ds1 = self.ds #downsample factor (2,2)
+        st0, st1 = self.st #stride size (2,2)
+        pd0, pd1 = self.padding #padding beyond the border
+        inc_pad = int(self.mode == 'average_inc_pad')
+        sum_mode = int(self.mode == 'sum')
+        skip_pad = int((self.mode == 'sum') or (self.mode == 'average_inc_pad'))
+        ccode = """
+        // sanity checks
+        int x_typenum = PyArray_ObjectType((PyObject*)%(x)s, 0);
+        int gz_typenum = PyArray_ObjectType((PyObject*)%(gz)s, 0);
+        
+        if (x_typenum != gz_typenum)
+        {
+            PyErr_SetString(PyExc_ValueError, "input types must match");
+            %(fail)s;
+        }
+        if(PyArray_NDIM(%(x)s)!=4)
+        {
+            PyErr_SetString(PyExc_ValueError, "x must be a 4d ndarray");
+            %(fail)s;
+        }
+        if(PyArray_NDIM(%(gz)s)!=4)
+        {
+            PyErr_SetString(PyExc_ValueError, "gz must be a 4d ndarray");
+            %(fail)s;
+        }       
+        
+        int r, c; // shape of the padded x input
+        r = PyArray_DIMS(%(x)s)[2];
+        c = PyArray_DIMS(%(x)s)[3];
+        r += %(pd0)s * 2;
+        c += %(pd1)s * 2;
+
+        int z_r, z_c; //shape of the (unsupplied) z input
+        if (%(pd0)s != 0 && %(pd1)s != 0 && !%(ignore_border)s)
+            {
+              PyErr_SetString(PyExc_ValueError,
+                "padding must be (0,0) when ignore border is False");
+              %(fail)s;
+            }
+        if (%(ignore_border)s)
+        {
+            // '/' in C is different from '/' in python
+            if (r - %(ds0)s < 0)
+            {
+              z_r = 0;
+            }
+            else
+            {
+              z_r = (r - %(ds0)s) / %(st0)s + 1;
+            }
+            if (c - %(ds1)s < 0)
+            {
+              z_c = 0;
+            }
+            else
+            {
+              z_c = (c - %(ds1)s) / %(st1)s + 1;
+            }
+        }
+        else
+        {
+            // decide how many rows the output has
+            if (%(st0)s >= %(ds0)s)
+            {
+                z_r = (r - 1) / %(st0)s + 1;
+            }
+            else
+            {
+                z_r = std::max(0, (r - 1 - %(ds0)s) / %(st0)s + 1) + 1;
+            }
+            // decide how many columns the output has
+            if (%(st1)s >= %(ds1)s)
+            {
+                z_c = (c - 1) / %(st1)s + 1;
+            }
+            else
+            {
+                z_c = std::max(0, (c - 1 - %(ds1)s) / %(st1)s + 1) + 1;
+            }
+        }
+        
+        // allocating memory for gx - same  size of x input but no padding
+        if ((!%(gx)s)
+          || !PyArray_ISCONTIGUOUS(%(gx)s)
+          || *PyArray_DIMS(%(gx)s)!=4
```

Comment:
Thanks Frederic. If you can get it to work feel free to amend and
incorporate. I won't be able to attend to this for another 2 or 3 weeks.

I note however that all of these operations are very similar to
convolutions. Is it worth looking at extending the convolution code? It
already has various optimisations and automated tweakers.
On 30/09/2015 2:48 PM, "Frédéric Bastien" <notifications@github.com> wrote:

> In theano/tensor/signal/downsample.py
> <https://github.com/Theano/Theano/pull/3422#discussion_r40788782>:
>
> > +            }
> > +            // decide how many columns the output has
> > +            if (%(st1)s >= %(ds1)s)
> > +            {
> > +                z_c = (c - 1) / %(st1)s + 1;
> > +            }
> > +            else
> > +            {
> > +                z_c = std::max(0, (c - 1 - %(ds1)s) / %(st1)s + 1) + 1;
> > +            }
> > +        }
> > +
> > +        // allocating memory for gx - same  size of x input but no padding
> > +        if ((!%(gx)s)
> > +          || !PyArray_ISCONTIGUOUS(%(gx)s)
> > +          || *PyArray_DIMS(%(gx)s)!=4
>
> There is a mistake here. It should be:
> PyArray_NDIM(%(gx)s)!=4
>
> We have a PR that fix this at other places (probably you took it from
> there)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3422/files#r40788782>.
>